Rejigger API Priority and Fairness config

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug

/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup
/kind bug

#### What this PR does / why we need it:
This PR tweaks the default configuration objects for API Priority and Fairness. There are two changes.

1. Introduce a FlowSchema and priority level for events.
2. Generally reshuffle the nominal concurrency shares and borrowing parameters so that the baseline is more in line with actual priority and need and can respond via borrowing. This is good because the default configuration was originally designed before borrowing was introduced and has some inversions among the nominal concurrency shares. See https://github.com/kubernetes/kubernetes/issues/121982#issuecomment-1933534198 for most of this. Also, @linxiulei has been recommending increasing the nominal concurrency shares for leader election, and testing supports this change. See #129646 and the other PRs and tests referenced in https://github.com/kubernetes/kubernetes/pull/129646#issuecomment-2649193226 .

This PR subsumes #129646 .

Also, as suggested in https://github.com/kubernetes/kubernetes/pull/129646#discussion_r1971451869 , this PR introduces a FeatureGate that allows administrators to opt out of this set of configuration changes. The FeatureGate is named "APFNewConifg".

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #121982

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The default configuration for API Priority and Fairness has changed, to work better under stress and take advantage of the borrowing of concurrency between priority levels. Note that during a rollout, while a cluster has a mixed population of kube-apiservers, they will periodically overwrite each other's configurations; this is expected to be a usable compromise. There is a new FeatureGate named "APFNewConifg" that can be disabled to opt out of the new configuration.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Increase leader-election nominal concurrency shares from 10 to 40

Because some testing shows leader election being starved.

Signed-off-by: Mike Spreitzer <mspreitz@us.ibm.com>

MikeSpreitzer committed:
Increase leader-election's lendable percent

.. so that it takes no larger bite from system capacity than before
(modulo small difference due to total nominal shares).

MikeSpreitzer committed:
Add FlowSchema and priority level for events

MikeSpreitzer committed:
Change APF config to stop working around lack of borrowing

MikeSpreitzer committed:
Introduce FeatureGate controlling which APF config to use

MikeSpreitzer commented:
/cc @wojtek-t 
/cc @tkashem 
/cc @deads2k 
/cc @linxiulei

wojtek-t commented on pkg/features/kube_features.go:
```diff
@@ -58,6 +58,16 @@ const (
 	AnyVolumeDataSource featuregate.Feature = "AnyVolumeDataSource"
 
+	// owner: @MikeSpreitzer, @tkashem, @linxiulei
+	// beta: 1.33
```
nit: we no longer put stage in the comments (it's reversible from the versioned feature list)

MikeSpreitzer replied:
removed.

wojtek-t commented on pkg/features/kube_features.go:
```diff
@@ -58,6 +58,16 @@ const (
+	// - introduce priority level and flow schema for events;
+	// - generally reorganize to stop working around lack of borrowing;
+	// - increase the nominal concurrency shares for leader election.
+	APFNewConfig featuregate.Feature = "APFNewConifg"
```
What if we will decide to change APF config again in the future? 
Maybe call it APFv133Config or sth?

MikeSpreitzer replied:
I assume that the form of growth will be additional features, since a feature is defined to be a boolean choice.

MikeSpreitzer replied:
I called it "APFv134Config", since release 1.33 enhancements are already frozen.

siyuanfoundation commented:
/triage accepted

MikeSpreitzer commented on pkg/apis/flowcontrol/internalbootstrap/default-internal.go:
```diff
@@ -19,23 +19,54 @@ package internalbootstrap
+
+// PrioritylevelConfigurations maps `newConfig bool` to a slice of all the default objects
+// Deeply immutable.
+var PrioritylevelConfigurations = map[bool][]*flowcontrolv1.PriorityLevelConfiguration{
```
Re 1. Would this mean two FeatureGates, thus four possible configs to use? I would not want to get this complex unless we really need it. Did you have a look at the studies referenced by https://github.com/kubernetes/kubernetes/pull/129646#issuecomment-2649193226 ? It looks to me like all the changes that we are talking about play nice together, and get the result that in @linxiulei's most stressful test it is the events that do almost all of the suffering.

MikeSpreitzer replied:
Re 2. I figured we would cross that bridge if and when we come to it. Hoping that we won't :-).

MikeSpreitzer replied:
The latest rev re-organizes the dependency on Feature(s), to better prepare for more such Features in the future. Now the main way to get config is to call a func whose arguments are the values of the relevant Features.

MikeSpreitzer replied:
.. and revised again, so that the funcs take one whole FeatureGate rather than a series of Feature values.

MikeSpreitzer commented:
The force-push to 2e962219988 is a rebase and fixing the oversights noted in review.

MikeSpreitzer commented on pkg/apis/flowcontrol/validation/validation.go:
```diff
@@ -381,12 +387,15 @@ func ValidateIfMandatoryPriorityLevelConfigurationObject(pl *flowcontrol.Priorit
 
+	// For now, accept either the old or the new default config.
+	// In a later release, accept only the new.
+	mandOld := internalbootstrap.MandatoryPriorityLevelConfigurations[false][pl.Name]
```
Fixed.

MikeSpreitzer commented on pkg/apis/flowcontrol/validation/validation.go:
```diff
@@ -83,13 +83,18 @@ func ValidateFlowSchema(fs *flowcontrol.FlowSchema) field.ErrorList {
+	if mand, ok := internalbootstrap.MandatoryFlowSchemas[true][fs.Name]; ok {
+		// The old config objects are a subset of the new config objects,
+		// as far as identify is concerned. The specs may differ.
+		mandOld := internalbootstrap.MandatoryFlowSchemas[false][fs.Name]
```
Fixed.

MikeSpreitzer commented:
The force-push to d97e207b0d1 renames the FeatureGate from "APFNewConfig" to "APFv134Config".

MikeSpreitzer commented:
The force-push to 8b9f4ddd028 is a rebase + add a commit that changes the way that the dependency on Features is handled, to better prepare for the possibility of more Features that affect APF configuration.

deads2k commented on pkg/apis/flowcontrol/validation/validation.go:
```diff
@@ -83,13 +83,18 @@ func ValidateFlowSchema(fs *flowcontrol.FlowSchema) field.ErrorList {
+	if mand, ok := internalbootstrap.GetMandatoryFlowSchemasMap(true)[fs.Name]; ok {
+		// The old config objects are a subset of the new config objects,
+		// as far as identify is concerned. The specs may differ.
+		mandOld := internalbootstrap.GetMandatoryFlowSchemasMap(false)[fs.Name]
```
if we assume as @wojtek-t does (and I think it's a good point), if we have more than two sets of flowschemas to consider, the `false` probably needs to be string, right?

MikeSpreitzer replied:
I was thinking that the arguments to that function will be the values of the relevant Features. As we add and remove relevant Features, the function's signature will change and the compiler will help us find all the call sites, so we know that we correctly updated them.

MikeSpreitzer replied:
Lately I have updated the PR to use a FeatureGate instead of the collection of relevant Feature values.

MikeSpreitzer commented:
The force-push to 8a94a799e9c is a rebase onto `upstream/master`.

MikeSpreitzer committed:
Add integration test of APF config flapping

Make APFv134Config Alpha at 1.33

MikeSpreitzer commented:
/retest

MikeSpreitzer commented:
I have a problem with the new integration test. The pull-kubernetes-verify job complained about the way that the integration test flaps the Feature, saying

```
 Test files may not access mutable global feature gates directly:
test/integration/apiserver/flowcontrol/concurrency_test.go:			case featuregate.MutableFeatureGate:
Use this invocation instead:
  featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>) 
```

But https://github.com/kubernetes/component-base/blob/v0.32.3/featuregate/testing/feature_gate.go#L40 says that it cannot flap one Feature.

MikeSpreitzer committed:
Use featuregatetesting.SetFeatureGateDuringTest

.. and use it multiple times on the same Feature in the same test.
The comment on this func says it will fail in that case, but it
appears to actually succeed. Examining the body of the func, I do
not see why it should fail.

MikeSpreitzer commented:
Erm, well, maybe that comment about not setting the same Feature twice in one test is just too cautious? I looked at the body of that func, and do not see a failure. I tested, and also did not get a failure. I have added a commit that switches to using that func.

MikeSpreitzer committed:
Use more canonical structure for multi-phase test

MikeSpreitzer committed:
Select APF config by FeatureGate

.. instead of feature values.

MikeSpreitzer commented on test/integration/apiserver/flowcontrol/config_flap_test.go:
```diff
@@ -0,0 +1,175 @@
+		return
+	}
+	tCtx.Cancel("test function done")
+}
```
Oh, of course. Done.

MikeSpreitzer commented:
The force push to a80c450b0e7 is a rebase onto `master`.

MikeSpreitzer commented:
And the force-push to e3e70fc413b fixes a botch in that rebase.

MikeSpreitzer commented:
The force-push to 58b961e08a0 changes a variable name to not trigger a sloppy linter (`hack/verify-non-mutating-validation.sh`).
